src/Plugin/search_api/processor/ProductStatus.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+namespace Drupal\commerce\Plugin\search_api\processor;
+
+use Drupal\commerce_product\Entity\ProductInterface;
+use Drupal\commerce_product\Entity\ProductVariationInterface;
+use Drupal\search_api\IndexInterface;
+use Drupal\search_api\Processor\ProcessorPluginBase;
+
+/**
+ * Excludes unpublished nodes from node indexes.
+ *
+ * @SearchApiProcessor(
+ *   id = "commerce_product_status",
+ *   label = @Translation("Product status"),
+ *   description = @Translation("Exclude unpublished products from being indexed."),
+ *   stages = {
+ *     "alter_items" = 0,
+ *   },
+ * )
+ */
+class ProductStatus extends ProcessorPluginBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function supportsIndex(IndexInterface $index) {
+    $supported_entity_types = ['commerce_product', 'commerce_product_variation'];
```

Comment:
bojanz said we should sync with the Search API maintainers about part of this. We should see if their processor could support all entity types which implement EntityPublishedInterface. We'll be there after #763.

However, we'd still need specific for variations. Unless variations implemented EntityPublishedInterface (it has status key.) And it's published status was based on its own value and parent product. In latter case we wouldn't need this.

- [ ] Reach out to Search API maintainers about patching the EntityStatus processor
- [ ] Discuss with @bojanz about Variations implementing EntityPublishedInterface

Reply:
I'd love to see us refocus this plugin to variations (commerce_product_variation_status) while fixing Search API to automatically handle entity status.

Reply:
Search API issue: https://www.drupal.org/node/2907756